Web: Making keyboard Navigation work for login/signin buttons.

CIC4DA opened the pull request:
It is a continuation/completion of the PR #28074, made by @sethdivyansh.
Fixes: #28037

<!-- Describe your pull request here.-->
 
 In this PR, We have refactored the ```navbar_menus.js``` with it's basic functionality of how it works, as before there was multiple duplicative code. Now it works on the basis of list system, and options can be visited with left and right press of arrow keys.
 
 Now after pressing ```G``` Hotkey, in the spectator mode, and then travelling with arrow keys, will take you to login/signup button also.
 
**Screenshots and screen captures:**

 **Steps can be seen in the video capture.**
                                      1. Changed focus to ```login button``` from ```Help menu``` on pressing ```left arrow key```.
                                      2. Changed focus from ```login button``` to ```signup button``` on pressing ```left arrow key```.
                                      3. Changed focus from ```signup button``` to ```login button``` on pressing ```right arrow key```.
                                      4. Changed focus from ```login button``` to ```Help menu``` on pressing ```right arrow key```.


**Dark**

https://github.com/zulip/zulip/assets/91906953/c0f6181e-f127-40eb-8299-b2d5129b2108


**Light**

https://github.com/zulip/zulip/assets/91906953/4f26f04d-f741-4052-99a2-77e5b6db8738


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

CIC4DA commented:
@alya I have updated the commit message to match the [commit style guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages).

alya commented:
Thanks! It looks like you also need to complete the self-review checklist.

CIC4DA commented:
@alya updated the self-review checklist.

alya commented:
@amanagr Could you please review this one? I haven't tested.

amanagr commented on web/src/navbar_menus.js:
```diff
@@ -4,54 +4,101 @@ import {page_params} from "./page_params";
-        personal_menu_popover.toggle();
-        gear_menu.toggle();
-        return true;
+    let i = 0;
+    for (const selected_popover of popover_menus_options) {
+        if (is_menu_displayed(selected_popover)) {
+            return handle_menu_event(selected_popover, event_name, i);
+        }
+        i = i + 1;
```
This can be writtten as:
```js
export function handle_keyboard_events(event_name) {
    const index = popover_menus_options.findIndex(is_menu_displayed);

    if (index !== -1) {
        return handle_menu_event(popover_menus_options[index], event_name, index);
    }

    return false;
}
```

CIC4DA replied:
Updated.

amanagr commented on web/src/navbar_menus.js:
```diff
@@ -4,54 +4,101 @@ import {page_params} from "./page_params";
-        popover_menus.is_personal_menu_popover_displayed() ||
-        popover_menus.is_gear_menu_popover_displayed() ||
-        popover_menus.is_help_menu_popover_displayed()
+        document.activeElement &&
```
`document.activeElement` should only matter for signup/login buttons. Let's keep `is_navbar_menus_displayed` the same and add another function which checks this named `can_handle_navigation_hotkey`.

CIC4DA replied:
Updated as Suggested.

amanagr replied:
I was thinking `can_handle_navigation_hotkey` will have the `|| is_navbar_menus_displayed` inside it.

amanagr commented on web/templates/navbar.hbs:
```diff
@@ -33,7 +33,7 @@
                 <a href="/register/"  class="signup_button">
                     {{t 'Sign up' }}
                 </a>
-                <a class="login_button">
+                <a href="/login/" class="login_button">
                     {{t 'Log in' }}
```
You can use `tabindex` instead of `href` here.

CIC4DA replied:
Updated

amanagr commented:
@CIC4DA looks great! Left a few comments.

alya commented:
@CIC4DA In addition to specific comments, it's helpful to post a summary comment when you've addressed all the feedback and your PR is ready for another round of review.

It looks like that's the case here?

CIC4DA commented:
@amanagr I have addressed the above feedback, could you please review it.

@alya Apologies for the delay, but there was one remaining change that needed to be made. I have now fully addressed all the feedback.

amanagr commented:
Oh! We forgot to handle it for narrow widths.

<img width="300" alt="Screenshot 2024-04-02 at 9 30 12 AM" src="https://github.com/zulip/zulip/assets/25124304/4cd18d76-fd08-47e4-a683-7ae843d6d82c">

amanagr commented:
You can use conditions like `window.innerWidth < media_breakpoints_num.xl` to check for narrow state.

alya commented:
Closing in favor of #29781. Thanks for the work here, @CIC4DA !
